File: Kotlin/BridgeCoordinator.kt
/**
 * W4RP Bridge - Coordinator
 *
 * Handles WebView message routing and coordinates between
 * JavaScript bridge and native BLE operations.
 *
 * @license MIT
 * @copyright 2024 W4RP Automotive
 * @version 2.0.0
 */

package com.w4rp.bridge

import android.webkit.JavascriptInterface
import android.webkit.WebView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject

/**
 * Coordinates communication between Android WebView and W4RPBridge.
 *
 * Responsibilities:
 * - Receives JavaScript messages via @JavascriptInterface
 * - Routes method calls to W4RPBridge
 * - Sends responses back to JavaScript
 * - Synchronizes state between native and JS
 *
 * Usage:
 * ```kotlin
 * val bridge = W4RPBridge(context)
 * val coordinator = W4RPCoordinator(bridge)
 * webView.addJavascriptInterface(coordinator, W4RPInjection.INTERFACE_NAME)
 * webView.evaluateJavascript(W4RPInjection.jsBridge, null)
 * ```
class W4RPBridgeCoordinator(private val bridge: W4RPBridge) {
    
    private var webView: WebView? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    
    /** Optional callback for QR code scanning. */
    var onScanQRCode: ((callback: (String?) -> Unit) -> Unit)? = null
    
    /**
     * Attach WebView and configure bridge callbacks.
     */
    fun attach(webView: WebView) {
        this.webView = webView
        configureBridgeCallbacks()
        pushCurrentState()
    }
    
    // =========================================================================
    // JavaScript Interface
    // =========================================================================
    
    /**
     * Called from JavaScript via W4RPBridgeNative.postMessage(json).
     */
    @JavascriptInterface
    fun postMessage(messageJson: String) {
        try {
            val json = JSONObject(messageJson)
            val id = json.optInt("id", 0)
            val method = json.optString("method", "")
            val params = json.optJSONObject("params") ?: JSONObject()
            
            // Route on main thread
            scope.launch {
                handleMethod(id, method, params)
            }
        } catch (e: Exception) {
            android.util.Log.e("W4RP", "Failed to parse message: $messageJson", e)
        }
    }
    
    // =========================================================================
    // Method Routing
    // =========================================================================
    
    private suspend fun handleMethod(id: Int, method: String, params: JSONObject) {
        when (method) {
            "scan" -> handleScan(id, params)
            "stopScan" -> {
                bridge.stopScan()
                sendSuccess(id)
            }
            "connect" -> handleConnect(id, params)
            "disconnect" -> handleDisconnect(id)
            "getProfile" -> handleGetProfile(id)
            "getRules" -> handleGetRules(id)
            "setRules" -> handleSetRules(id, params)
            "setDebugMode" -> handleSetDebugMode(id, params)
            "watchDebugSignals" -> handleWatchDebugSignals(id, params)
            "startOTA" -> handleStartOTA(id, params)
            "stopOTA" -> handleStopOTA(id)
            "scanQRCode" -> handleScanQRCode(id)
            "appReady" -> handleAppReady(id)
            else -> sendError(id, 9999, "Unknown method: $method")
        }
    }
    
    // =========================================================================
    // Handlers
    // =========================================================================
    
    private suspend fun handleScan(id: Int, params: JSONObject) {
        val timeoutMs = params.optLong("timeoutMs", 8000)
        
        try {
            val devices = bridge.scan(timeoutMs)
            val devicesArray = JSONArray()
            devices.forEach { device ->
                devicesArray.put(JSONObject(device.toMap()))
            }
            sendSuccess(id, devicesArray)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "Scan failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.CONNECTION_FAILED.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleConnect(id: Int, params: JSONObject) {
        val deviceId = params.optString("deviceId", "")
        val device = bridge.discoveredDevices.find { it.device.address == deviceId }
        
        if (device == null) {
            sendError(id, W4RPErrorCode.DEVICE_NOT_FOUND.code, "Device not found")
            return
        }
        
        try {
            bridge.connect(device)
            updateConnectedDevice(device)
            sendSuccess(id)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "Connection failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.CONNECTION_FAILED.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleDisconnect(id: Int) {
        try {
            bridge.disconnect()
            updateConnectedDevice(null)
            sendSuccess(id)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "Disconnect failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.CONNECTION_FAILED.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleGetProfile(id: Int) {
        try {
            // Get raw binary from bridge
            val data = bridge.getProfile()
            
            // Send as base64 - UI will parse
            val base64 = android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)
            sendSuccess(id, base64)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "getProfile failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.INVALID_DATA.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleGetRules(id: Int) {
        try {
            // Get raw binary from bridge
            val data = bridge.getRules()
            
            // Send as base64 - UI will parse
            val base64 = android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)
            sendSuccess(id, base64)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "getRules failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.INVALID_DATA.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleSetRules(id: Int, params: JSONObject) {
        // WBP v2: Expect base64-encoded binary
        val base64 = params.optString("binary", "")
        
        if (base64.isNotEmpty()) {
            try {
                val binary = android.util.Base64.decode(base64, android.util.Base64.DEFAULT)
                val persistent = params.optBoolean("persistent", true)
                
                bridge.setRules(binary, persistent) { sent, total ->
                    scope.launch {
                        evaluateJS("window.W4RPBridge._notifyProgress($sent, $total);")
                    }
                }
                sendSuccess(id)
            } catch (e: W4RPError) {
                sendError(id, e.code.code, e.message ?: "setRules failed")
            } catch (e: Exception) {
                sendError(id, W4RPErrorCode.WRITE_FAILED.code, e.message ?: "Unknown error")
            }
            return
        }
        
        sendError(id, W4RPErrorCode.INVALID_DATA.code, "Missing binary parameter (base64)")
    }
    
    private suspend fun handleSetDebugMode(id: Int, params: JSONObject) {
        val enabled = params.optBoolean("enabled", false)
        
        try {
            bridge.setDebugMode(enabled)
            sendSuccess(id)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "setDebugMode failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.WRITE_FAILED.code, e.message ?: "Unknown error")
        }
    }
    
    private suspend fun handleWatchDebugSignals(id: Int, params: JSONObject) {
        val signalsArray = params.optJSONArray("signals")
        
        if (signalsArray == null) {
            sendError(id, W4RPErrorCode.INVALID_DATA.code, "Missing signals parameter")
            return
        }
        
        // Parse as array of signal objects (not strings)
        val signals = (0 until signalsArray.length()).map { i ->
            val obj = signalsArray.getJSONObject(i)
            obj.keys().asSequence().associateWith { key -> obj.get(key) as Any }
        }
        
        try {
            bridge.watchDebugSignals(signals)
            sendSuccess(id)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "watchDebugSignals failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.WRITE_FAILED.code, e.message ?: "Unknown error")
        }
    }
    
    private fun handleStopOTA(id: Int) {
        bridge.stopOTA()
        sendSuccess(id)
    }
    
    private suspend fun handleStartOTA(id: Int, params: JSONObject) {
        val patchBase64 = params.optString("patch", "")
        
        if (patchBase64.isEmpty()) {
            sendError(id, W4RPErrorCode.INVALID_DATA.code, "Invalid patch data")
            return
        }
        
        try {
            val patchData = android.util.Base64.decode(patchBase64, android.util.Base64.DEFAULT)
            bridge.startOTA(patchData)
            sendSuccess(id)
        } catch (e: W4RPError) {
            sendError(id, e.code.code, e.message ?: "OTA failed")
        } catch (e: Exception) {
            sendError(id, W4RPErrorCode.INVALID_DATA.code, e.message ?: "Unknown error")
        }
    }
    
    private fun handleScanQRCode(id: Int) {
        val handler = onScanQRCode
        if (handler == null) {
            sendError(id, 9999, "QR scanning not available")
            return
        }
        
        handler { code ->
            if (code != null) {
                sendSuccess(id, code)
            } else {
                sendError(id, 1103, "QR scan cancelled")
            }
        }
    }
    
    private fun handleAppReady(id: Int) {
        pushCurrentState()
        sendSuccess(id)
    }
    
    // =========================================================================
    // Bridge Callbacks
    // =========================================================================
    
    private fun configureBridgeCallbacks() {
        bridge.onConnectionStateChanged = { state ->
            updateConnectionState(state)
        }
        
        bridge.onBluetoothStatusChanged = { status ->
            pushBluetoothStatus(status)
        }
        
        bridge.onDeviceDiscovered = { devices ->
            val devicesArray = JSONArray()
            devices.forEach { device ->
                devicesArray.put(JSONObject(device.toMap()))
            }
            evaluateJS("window.W4RPBridge._notifyDeviceDiscovered($devicesArray);")
        }
        
        bridge.onStatusUpdate = { json ->
            evaluateJS("window.W4RPBridge._notifyStatus($json);")
        }
        
        bridge.onDebugData = { data ->
            val dataJson = JSONObject().apply {
                put("id", data.id)
                put("type", data.type.value)
                data.value?.let { put("value", it) }
                data.active?.let { put("active", it) }
            }
            evaluateJS("window.W4RPBridge._notifyDebug($dataJson);")
        }
        
        bridge.onReconnecting = { attempt ->
            evaluateJS("window.W4RPBridge._notifyReconnecting($attempt);")
        }
        
        bridge.onReconnected = {
            evaluateJS("window.W4RPBridge._notifyReconnected();")
        }
        
        bridge.onReconnectFailed = { error ->
            val errorJson = JSONObject().apply {
                put("code", error.code.code)
                put("message", error.message)
            }
            evaluateJS("window.W4RPBridge._notifyReconnectFailed($errorJson);")
        }
        
        bridge.onError = { error ->
            val errorJson = JSONObject().apply {
                put("code", error.code.code)
                put("message", error.message)
                error.context?.let { put("context", JSONObject(it)) }
            }
            evaluateJS("window.W4RPBridge._notifyError($errorJson);")
        }
    }
    
    // =========================================================================
    // JS Communication
    // =========================================================================
    
    private fun sendSuccess(id: Int, result: Any? = null) {
        val resultStr = when (result) {
            null -> "null"
            is JSONObject, is JSONArray -> result.toString()
            else -> result.toString()
        }
        evaluateJS("window.W4RPBridge._resolve($id, $resultStr);")
    }
    
    private fun sendError(id: Int, code: Int, message: String) {
        val errorJson = """{"code":$code,"message":"${message.replace("\"", "\\\"")}"}"""
        evaluateJS("window.W4RPBridge._reject($id, $errorJson);")
    }
    
    private fun updateConnectionState(state: W4RPConnectionState) {
        evaluateJS("window.W4RPBridge._setConnectionState('${state.name}');")
    }
    
    fun pushBluetoothStatus(status: W4RPBluetoothStatus? = null) {
        val s = status ?: bridge.bluetoothStatus.value
        evaluateJS("window.W4RPBridge._setBluetoothStatus('${s.name}');")
    }
    
    private fun updateConnectedDevice(device: W4RPDevice?) {
        if (device != null) {
            val deviceJson = JSONObject(device.toMap()).toString()
            evaluateJS("window.W4RPBridge._setConnectedDevice($deviceJson);")
        } else {
            evaluateJS("window.W4RPBridge._setConnectedDevice(null);")
        }
    }
    
    private fun pushCurrentState() {
        scope.launch {
            updateConnectionState(bridge.connectionState)
            pushBluetoothStatus(bridge.bluetoothStatus.value)
            bridge.connectedDevice?.let { updateConnectedDevice(it) }
        }
    }
    
    private fun evaluateJS(script: String) {
        webView?.post {
            webView?.evaluateJavascript(script, null)
        }
    }
}

File: Swift/BridgeCoordinator.swift
/**
 * W4RP Bridge - Coordinator
 *
 * Handles WKWebView message routing and coordinates between
 * JavaScript bridge and native BLE operations.
 *
 * @license MIT
 * @copyright 2024 W4RP Automotive
 * @version 2.0.0
 */

import Foundation
import WebKit

// MARK: - W4RPBridgeCoordinator

/**
 * Coordinates communication between WKWebView and W4RPBridge.
 *
 * Responsibilities:
 * - Receives JavaScript messages via WKScriptMessageHandler
 * - Routes method calls to W4RPBridge
 * - Sends responses back to JavaScript
 * - Synchronizes state between native and JS
 */
@MainActor
public final class W4RPBridgeCoordinator: NSObject, WKScriptMessageHandler {
    
    // MARK: - Properties
    
    public let bridge: W4RPBridge
    public weak var webView: WKWebView? {
        didSet {
            if webView != nil {
                pushCurrentState()
            }
        }
    }
    
    /** Optional callback for QR code scanning. */
    public var onScanQRCode: ((@escaping (String?) -> Void) -> Void)?
    
    // MARK: - Init
    
    public init(bridge: W4RPBridge) {
        self.bridge = bridge
        super.init()
        print("🔵 [Coordinator] Init - bridge.bluetoothStatus = \(bridge.bluetoothStatus.rawValue)")
        configureBridgeCallbacks()
    }
    
    // MARK: - WKScriptMessageHandler
    
    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == W4RPBridge.messageHandlerName,
              let body = message.body as? [String: Any] else { return }
        
        let id = body["id"] as? Int ?? 0
        let method = body["method"] as? String ?? ""
        let params = body["params"] as? [String: Any] ?? [:]
        
        handleMethod(id: id, method: method, params: params)
    }
    
    // MARK: - Method Routing
    
    private func handleMethod(id: Int, method: String, params: [String: Any]) {
        switch method {
        case "scan":
            handleScan(id: id, params: params)
        case "stopScan":
            bridge.stopScan()
            sendSuccess(id: id)
        case "connect":
            handleConnect(id: id, params: params)
        case "disconnect":
            handleDisconnect(id: id)
        case "getProfile":
            handleGetProfile(id: id)
        case "getRules":
            handleGetRules(id: id)
        case "setRules":
            handleSetRules(id: id, params: params)
        case "setDebugMode":
            handleSetDebugMode(id: id, params: params)
        case "watchDebugSignals":
            handleWatchDebugSignals(id: id, params: params)
        case "startOTA":
            handleStartOTA(id: id, params: params)
        case "stopOTA":
            handleStopOTA(id: id)
        case "scanQRCode":
            handleScanQRCode(id: id)
        case "appReady":
            handleAppReady(id: id)
        default:
            sendError(id: id, code: 9999, message: "Unknown method: \(method)")
        }
    }
    
    // MARK: - Handlers
    
    private func handleScan(id: Int, params: [String: Any]) {
        let timeoutMs = params["timeoutMs"] as? Int ?? 8000
        let timeout = TimeInterval(timeoutMs) / 1000.0
        
        // State will be updated via bridge.onConnectionStateChange callback
        
        Task {
            do {
                let devices = try await bridge.scan(timeout: timeout)
                // Do NOT manually update state - let BLE callbacks handle it
                sendSuccess(id: id, result: devices.map { $0.toDict() })
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 1000, message: "\(error)")
            }
        }
    }
    
    private func handleConnect(id: Int, params: [String: Any]) {
        guard let deviceId = params["deviceId"] as? String,
              let uuid = UUID(uuidString: deviceId),
              let device = bridge.discoveredDevices.first(where: { $0.id == uuid }) else {
            sendError(id: id, code: W4RPErrorCode.deviceNotFound.rawValue, message: "Device not found")
            return
        }
        
        // State will be updated via bridge.onConnectionStateChange callback
        
        Task {
            do {
                try await bridge.connect(to: device)
                // Do NOT manually update state - let BLE callbacks handle it
                updateConnectedDevice(device)
                sendSuccess(id: id)
            } catch let error as W4RPError {
                // Do NOT manually update state - let BLE callbacks handle it
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 1000, message: "\(error)")
            }
        }
    }
    
    private func handleDisconnect(id: Int) {
        Task {
            do {
                try await bridge.disconnect()
                // Do NOT manually update state - let BLE callbacks handle it
                updateConnectedDevice(nil)
                sendSuccess(id: id)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 1000, message: "\(error)")
            }
        }
    }
    
    private func handleGetProfile(id: Int) {
        Task {
            do {
                // Get raw binary from bridge
                let data = try await bridge.getProfile()
                
                // Send as base64 - UI will parse
                let base64 = data.base64EncodedString()
                sendSuccess(id: id, result: base64)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 3000, message: "\(error)")
            }
        }
    }
    
    private func handleGetRules(id: Int) {
        Task {
            do {
                // Get raw binary from bridge
                let data = try await bridge.getRules()
                
                // Send as base64 - UI will parse
                let base64 = data.base64EncodedString()
                sendSuccess(id: id, result: base64)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 3000, message: "\(error)")
            }
        }
    }
    
    private func handleSetRules(id: Int, params: [String: Any]) {
        print("🔵 [Coordinator] handleSetRules params keys: \(params.keys)")
        
        // WBP v2: Expect base64-encoded binary
        if let base64 = params["binary"] as? String,
           let binary = Data(base64Encoded: base64) {
            print("🔵 [Coordinator] handleSetRules binary length: \(binary.count)")
            let persistent = params["persistent"] as? Bool ?? true
            
            Task {
                do {
                    try await bridge.setRules(binary: binary, persistent: persistent) { sent, total in
                        self.evaluateJS("window.W4RPBridge._notifyProgress(\(sent), \(total));")
                    }
                    sendSuccess(id: id)
                } catch let error as W4RPError {
                    sendError(id: id, code: error.code.rawValue, message: error.message)
                } catch {
                    sendError(id: id, code: 2000, message: "\(error)")
                }
            }
            return
        }
        
        print("🔴 [Coordinator] handleSetRules: missing 'binary' param (base64). Params: \(params)")
        sendError(id: id, code: W4RPErrorCode.invalidData.rawValue, message: "Missing binary parameter (base64)")
    }
    
    private func handleSetDebugMode(id: Int, params: [String: Any]) {
        let enabled = params["enabled"] as? Bool ?? false
        
        Task {
            do {
                try await bridge.setDebugMode(enabled: enabled)
                sendSuccess(id: id)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 2000, message: "\(error)")
            }
        }
    }
    
    private func handleWatchDebugSignals(id: Int, params: [String: Any]) {
        guard let signals = params["signals"] as? [[String: Any]] else {
            sendError(id: id, code: W4RPErrorCode.invalidData.rawValue, message: "Missing or invalid signals parameter")
            return
        }
        
        Task {
            do {
                try await bridge.watchDebugSignals(signals)
                sendSuccess(id: id)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 2000, message: "\(error)")
            }
        }
    }
    
    private func handleStopOTA(id: Int) {
        bridge.stopOTA()
        sendSuccess(id: id)
    }
    
    private func handleStartOTA(id: Int, params: [String: Any]) {
        guard let patchBase64 = params["patch"] as? String,
              let patchData = Data(base64Encoded: patchBase64) else {
            sendError(id: id, code: W4RPErrorCode.invalidData.rawValue, message: "Invalid patch data")
            return
        }
        
        Task {
            do {
                try await bridge.startOTA(patchData: patchData)
                sendSuccess(id: id)
            } catch let error as W4RPError {
                sendError(id: id, code: error.code.rawValue, message: error.message)
            } catch {
                sendError(id: id, code: 2000, message: "\(error)")
            }
        }
    }
    
    private func handleScanQRCode(id: Int) {
        guard let handler = onScanQRCode else {
            sendError(id: id, code: 9999, message: "QR scanning not available")
            return
        }
        
        handler { [weak self] code in
            if let code = code {
                self?.sendSuccess(id: id, result: code)
            } else {
                self?.sendError(id: id, code: 1103, message: "QR scan cancelled")
            }
        }
    }
    
    private func handleAppReady(id: Int) {
        pushCurrentState()
        sendSuccess(id: id)
    }
    
    // MARK: - Bridge Callbacks
    
    private func configureBridgeCallbacks() {
        bridge.onConnectionStateChanged = { [weak self] state in
            self?.updateConnectionState(state)
        }
        
        bridge.onDeviceDiscovered = { [weak self] devices in
            self?.notifyDevicesDiscovered(devices)
        }
        
        bridge.onStatusUpdate = { [weak self] json in
            self?.evaluateJS("window.W4RPBridge._notifyStatus('\(json.replacingOccurrences(of: "'", with: "\\'"))');")
        }
        
        bridge.onDebugData = { [weak self] data in
            guard let json = try? JSONEncoder().encode(data),
                  let jsonStr = String(data: json, encoding: .utf8) else { return }
            self?.evaluateJS("window.W4RPBridge._notifyDebug(\(jsonStr));")
        }
        
        bridge.onReconnecting = { [weak self] attempt in
            self?.evaluateJS("window.W4RPBridge._notifyReconnecting(\(attempt));")
        }
        
        bridge.onReconnected = { [weak self] in
            self?.evaluateJS("window.W4RPBridge._notifyReconnected();")
        }
        
        bridge.onReconnectFailed = { [weak self] error in
            self?.evaluateJS("window.W4RPBridge._notifyReconnectFailed({code:\(error.code.rawValue),message:'\(error.message)'});")
        }
        
        bridge.onBluetoothStatusChanged = { [weak self] status in
            print("🔵 [Coordinator] onBluetoothStatusChanged callback: \(status.rawValue)")
            self?.evaluateJS("window.W4RPBridge._setBluetoothStatus('\(status.rawValue)');")
        }
    }
    
    // MARK: - Response Helpers
    
    private func sendSuccess(id: Int, result: Any? = nil) {
        var response: [String: Any] = ["id": id, "success": true]
        if let result = result {
            response["result"] = result
        }
        
        if let data = try? JSONSerialization.data(withJSONObject: response),
           let json = String(data: data, encoding: .utf8) {
            evaluateJS("window.W4RPBridge._resolve(\(id), \(json).result);")
        }
    }
    
    private func sendError(id: Int, code: Int, message: String) {
        evaluateJS("window.W4RPBridge._reject(\(id), {code:\(code),message:'\(message.replacingOccurrences(of: "'", with: "\\'"))'});")
    }
    
    // MARK: - State Sync
    
    private func updateConnectionState(_ state: W4RPConnectionState) {
        evaluateJS("window.W4RPBridge._setConnectionState('\(state.rawValue)');")
    }
    
    private func updateConnectedDevice(_ device: W4RPDevice?) {
        if let device = device {
            print("🔵 [Coordinator] updateConnectedDevice: \(device.toDict())")
            if let data = try? JSONSerialization.data(withJSONObject: device.toDict()),
               let json = String(data: data, encoding: .utf8) {
                print("🔵 [Coordinator] Sending connectedDevice JSON: \(json)")
                evaluateJS("window.W4RPBridge._setConnectedDevice(\(json));")
            }
        } else {
            evaluateJS("window.W4RPBridge._setConnectedDevice(null);")
        }
    }
    
    private func notifyDevicesDiscovered(_ devices: [W4RPDevice]) {
        let list = devices.map { $0.toDict() }
        print("🔵 [Coordinator] notifyDevicesDiscovered: \(list)")
        if let data = try? JSONSerialization.data(withJSONObject: list),
           let json = String(data: data, encoding: .utf8) {
            print("🔵 [Coordinator] Sending devices JSON: \(json)")
            evaluateJS("window.W4RPBridge._notifyDeviceDiscovered(\(json));")
        }
    }
    
    private func pushCurrentState() {
        print("🔵 [Coordinator] pushCurrentState called - bluetoothStatus = \(bridge.bluetoothStatus.rawValue)")
        updateConnectionState(bridge.connectionState)
        if let device = bridge.connectedDevice {
            updateConnectedDevice(device)
        }
        pushBluetoothStatusInternal()
    }
    
    public func pushBluetoothStatus() {
        print("🔵 [Coordinator] pushBluetoothStatus (public) called - bluetoothStatus = \(bridge.bluetoothStatus.rawValue)")
        pushBluetoothStatusInternal()
    }
    
    private func pushBluetoothStatusInternal() {
        let js = "window.W4RPBridge._setBluetoothStatus('\(bridge.bluetoothStatus.rawValue)');"
        print("🔵 [Coordinator] Evaluating: \(js)")
        evaluateJS(js)
    }
    
    private func evaluateJS(_ script: String) {
        webView?.evaluateJavaScript(script, completionHandler: nil)
    }
}

public protocol EquatableBytes: Equatable {
    init(bytes: [UInt8])
    var bytes: [UInt8] { get }
}
